zap/src/main/java/org/zaproxy/zap/session/ScriptBasedSessionManagementMethodType.java

```diff
@@ -0,0 +1,668 @@
+                if (script == null) {
+                    log.error(
+                            "Unable to find script while loading Script Based Session Management Method for name: "
+                                    + scriptName);
+                    throw new ApiException(ApiException.Type.SCRIPT_NOT_FOUND, scriptName);
+                } else log.info("Loaded script for API:" + script.getName());
+                method.script = script;
+
+                SessionScript sessionScript = getScriptInterface(script);
+                String[] requiredParams = sessionScript.getRequiredParamsNames();
+                String[] optionalParams = sessionScript.getOptionalParamsNames();
+                if (log.isDebugEnabled()) {
+                    log.debug(
+                            "Loaded session management script - required parameters: "
+                                    + Arrays.toString(requiredParams)
+                                    + " - optional parameters: "
+                                    + Arrays.toString(optionalParams));
+                }
+
+                Map<String, String> paramValues = new HashMap<String, String>();
+                for (String rp : requiredParams) {
+                    // If one of the required parameters is not present, it will throw
+                    // an exception
+                    String val = ApiUtils.getNonEmptyStringParam(params, rp);
+                    paramValues.put(rp, val);
+                }
+
+                for (String op : optionalParams)
+                    paramValues.put(op, ApiUtils.getOptionalStringParam(params, op));
+                method.paramValues = paramValues;
+                if (log.isDebugEnabled())
+                    log.debug("Loaded session management script parameters:" + paramValues);
+
+                context.setSessionManagementMethod(method);
+            }
+        };
+    }
+
+    private ExtensionScript getScriptsExtension() {
+        if (extensionScript == null)
+            extensionScript =
+                    Control.getSingleton().getExtensionLoader().getExtension(ExtensionScript.class);
+        return extensionScript;
+    }
+
+    private SessionScript getScriptInterface(ScriptWrapper script) {
+        try {
+            return getScriptsExtension().getInterface(script, SessionScript.class);
+        } catch (Exception e) {
+            getScriptsExtension()
+                    .handleFailedScriptInterface(
+                            script,
+                            Constant.messages.getString(
+                                    "session.method.script.dialog.error.text.interface",
+                                    script.getName()));
+        }
+        return null;
+    }
+
+    /** The Interface that needs to be implemented by a Session Script. */
+    public interface SessionScript {
+
+        public ScriptBasedSession extractWebSession(HttpMessage msg);
```

Comment:
- Since it's expected a new session always it could be passed as parameter, instead of having the scripts create and return it. Maybe we should use a parameter object to allow to add more data/behaviour without breaking binary compatibility (same for other methods, clear/process)? We also need to pass the parameters.
 - The public modifier could be removed (same for other methods).